File: src/LeetCode/ThreeSum.java
package LeetCode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new LinkedList<>();
        if(nums.length <= 2){
            return result;
        }
        Arrays.sort(nums);


        for (int i = 0; i < nums.length - 2; i++ ){
            if (i == 0 || (i > 0 && nums[i] != nums[i-1])){

                int low = i+1;
                int high = nums.length - 1;
                int sum = -nums[i];
                while(high > low){
                    if (nums[high] + nums[low] == sum){
                        result.add(Arrays.asList(nums[i], nums[low], nums[high]));
                        while(high > low && nums[low] == nums[low + 1])low++;
                        while(high > low && nums[high] == nums[high - 1])high--;
                        low++;
                        high--;
                    } else if (nums[high] + nums[low] > sum){
                        high --;
                    } else {
                        low++;
                    }
                }
            }
        }
        return result;
    }
}

File: algoC/vectors/threesum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<int>> threesum(vector<int> arr, int sum)
{

    sort(arr.begin(), arr.end());
    int n = arr.size();
    vector<vector<int>> result;

    for (int i = 0; i <= n - 3; i++)
    {
        int j = i + 1;
        int k = n - 1;

        while (j < k)
        {
            int current_sum = arr[i] + arr[j] + arr[k];

            if (current_sum == sum)
            {
                result.push_back({arr[i], arr[j], arr[k]});
                j++;
                k--;
            }
            else if (current_sum > sum)
            {
                k--;
            }
            else
            {
                j++;
            }
        }
    }

    return result;
}

int main()
{

    std::vector<int> arr{1, 2, 3, 4, 5, 6, 7, 8, 9, 15};
    int S = 18;
    auto results = threesum(arr, S);

    cout << "s"
         << results.size() << endl;
    for (auto v : results)
    {
        for (auto s : v)
        {
            cout << s << ",";
        }
        cout << endl;
    }

    return 0;
}